Don't IdleExit the GPU Process if DOM Rendering is enabled and there is a client WebContent process

smfr opened the pull request:
#### 3db4e668e80eab852778350c423b9ed2f1b4ea06
<pre>
Don&apos;t IdleExit the GPU Process if DOM Rendering is enabled and there is a client WebContent process
<a href="https://bugs.webkit.org/show_bug.cgi?id=258487">https://bugs.webkit.org/show_bug.cgi?id=258487</a>
rdar://110742027

Reviewed by Wenson Hsieh.

Pass `useGPUProcessForDOMRenderingEnabled` in GPUProcessConnectionParameters, and keep
this flag up to date whenever pages are added or removed from the WebProcessProxy.
When this flag is true, return false from GPUConnectionToWebProcess::allowsExitUnderMemoryPressure().

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::allowsExitUnderMemoryPressure const):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/Shared/GPUProcessConnectionParameters.h:
* Source/WebKit/Shared/GPUProcessConnectionParameters.serialization.in:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createGPUProcessConnection):

Canonical link: <a href="https://commits.webkit.org/265623@main">https://commits.webkit.org/265623@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b374d8fc1ee567c5ddaa77a8781ea41152eb244

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11385 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11590 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11959 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13030 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10837 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13972 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11572 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13756 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11547 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12418 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9654 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13449 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9710 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17499 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10781 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10497 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13684 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10890 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8962 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10063 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2742 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14336 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10745 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -511,6 +511,9 @@ void WebProcessPool::createGPUProcessConnection(WebProcessProxy& webProcessProxy
     parameters.presentingApplicationAuditToken = configuration().presentingApplicationProcessToken();
 #endif
 
+    parameters.useGPUProcessForDOMRenderingEnabled = WTF::anyOf(webProcessProxy.pages(), [](const auto& page) {
```
I am a little worried we may create the GPUProcessConnection too eagerly and not have any pages yet. In this case, we set the flag to false and never update it. Shouldn't we send an updated flag to the GPUProcess whenever a page is added / removed from the WebProcess to be safe?

smfr replied:
That's a good point.

cdumez committed:
Don't IdleExit the GPU Process if DOM Rendering is enabled and there is a client WebContent process
https://bugs.webkit.org/show_bug.cgi?id=258487
rdar://110742027

Reviewed by Wenson Hsieh.

Pass `useGPUProcessForDOMRenderingEnabled` in GPUProcessConnectionParameters, and keep
this flag up to date whenever pages are added or removed from the WebProcessProxy.
When this flag is true, return false from GPUConnectionToWebProcess::allowsExitUnderMemoryPressure().

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::allowsExitUnderMemoryPressure const):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/Shared/GPUProcessConnectionParameters.h:
* Source/WebKit/Shared/GPUProcessConnectionParameters.serialization.in:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createGPUProcessConnection):

Canonical link: https://commits.webkit.org/265623@main
